File: src/backend/src/pke_backend/crypto/signatures.py
"""ECDSA P-256 signature verification with strict raw P1363 inputs.

The protocol pins ECDSA over P-256 with SHA-256 and a raw P1363 wire format
(exactly 64 bytes: `r || s`, each big-endian, left-padded to 32 bytes). DER
inputs and any other length must be rejected at this boundary — see
`context/16_canonical_encoding.md` (ECDSA section).

pyca/cryptography only accepts DER-encoded signatures through
`EllipticCurvePublicKey.verify`, so the wrapper:

1. Validates the wire length is exactly 64 bytes,
2. Splits into `(r, s)` and re-encodes via `encode_dss_signature` to feed
   the underlying primitive,
3. Maps `InvalidSignature` to `SignatureVerificationError` and any
   structural problem (length, type) to `SignatureFormatError` *before*
   the verify math runs.

`sign_payload` remains a stub: the backend is a verifier in this story and
does not produce signatures.
"""

from __future__ import annotations

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.utils import encode_dss_signature

from pke_backend.crypto.errors import SignatureFormatError, SignatureVerificationError

__all__ = ["sign_payload", "verify_payload", "verify_signature"]

_P1363_LENGTH = 64
_COORD_LENGTH = 32


def verify_signature(
    public_key: ec.EllipticCurvePublicKey,
    payload: bytes,
    sig: bytes,
) -> None:
    """Verify a raw P1363 ECDSA-P256-SHA256 signature over ``payload``.

    Returns ``None`` on success. Raises ``SignatureFormatError`` for any
    structural problem (wrong type, wrong length) before any verification
    math runs. Raises ``SignatureVerificationError`` when the signature is
    well-formed but does not validate.
    """
    if not isinstance(sig, (bytes, bytearray)):
        raise SignatureFormatError(reason=f"signature must be bytes, got {type(sig).__name__}")
    if len(sig) != _P1363_LENGTH:
        raise SignatureFormatError(reason=f"expected {_P1363_LENGTH}-byte raw P1363 signature, got {len(sig)} bytes")
    if not isinstance(public_key, ec.EllipticCurvePublicKey):
        raise SignatureFormatError(reason=f"public_key must be EllipticCurvePublicKey, got {type(public_key).__name__}")
    if not isinstance(public_key.curve, ec.SECP256R1):
        raise SignatureFormatError(reason=f"expected P-256 key, got curve {public_key.curve.name}")
    if not isinstance(payload, (bytes, bytearray)):
        raise SignatureFormatError(reason=f"payload must be bytes, got {type(payload).__name__}")

    r = int.from_bytes(bytes(sig[:_COORD_LENGTH]), "big")
    s = int.from_bytes(bytes(sig[_COORD_LENGTH:]), "big")
    der = encode_dss_signature(r, s)
    try:
        public_key.verify(der, bytes(payload), ec.ECDSA(hashes.SHA256()))
    except InvalidSignature as exc:
        raise SignatureVerificationError(reason="signature did not validate") from exc


def sign_payload(canonical: bytes, private_key: object) -> bytes:
    raise NotImplementedError


def verify_payload(canonical: bytes, signature: bytes, public_key: object) -> None:
    raise NotImplementedError

File: src/ios/PKE/Services/Crypto/Signatures.swift
// ECDSA P-256 sign / verify with strict raw-P1363 (64-byte) wire format.
// Mirrors the backend `pke_backend.crypto.signatures` contract from HLAM-16 and
// the canonical-encoding rule pinned in HLAM-3 / HLAM-25.

import Crypto
import Foundation

public enum Signatures {

    /// Sign `payload` with the supplied P-256 private key.
    /// Returns exactly 64 bytes (raw P1363 — `r || s`, each big-endian, left-padded to 32).
    /// Empty payload is permitted.
    public static func sign(
        payload: Data,
        with privateKey: P256.Signing.PrivateKey
    ) throws -> Data {
        do {
            let signature = try privateKey.signature(for: payload)
            return signature.rawRepresentation
        } catch {
            throw CryptoError.signatureFormat(reason: "sign failed")
        }
    }

    /// Verify a 64-byte raw P1363 `signature` over `payload` under `publicKey`.
    ///
    /// Order of checks (deliberate — the length gate runs before any parsing or verify math
    /// so attackers learn nothing about the key or digest from rejection latency):
    ///   1. `signature.count == 64`, else `.signatureFormat`.
    ///   2. parse into `ECDSASignature(rawRepresentation:)`, else `.signatureFormat`.
    ///   3. `publicKey.isValidSignature`, else `.signatureVerification`.
    public static func verify(
        _ signature: Data,
        of payload: Data,
        by publicKey: P256.Signing.PublicKey
    ) throws {
        guard signature.count == 64 else {
            throw CryptoError.signatureFormat(
                reason: "expected 64-byte raw P1363, got \(signature.count) bytes"
            )
        }
        let parsed: P256.Signing.ECDSASignature
        do {
            parsed = try P256.Signing.ECDSASignature(rawRepresentation: signature)
        } catch {
            throw CryptoError.signatureFormat(reason: "rawRepresentation rejected")
        }
        let ok = publicKey.isValidSignature(parsed, for: payload)
        if !ok {
            throw CryptoError.signatureVerification(reason: "verify failed")
        }
    }
}
